lms/product/factory.py

```diff
@@ -0,0 +1,39 @@
+# We need to import all classes so we are aware of them for the mapping below
+from lms.product.blackboard import BlackboardProduct  # pylint: disable=unused-import
+from lms.product.canvas import CanvasProduct  # pylint: disable=unused-import
+from lms.product.generic import GenericProduct
+from lms.product.product import Product
+
+
+def get_product_from_request(request) -> Product:
+    """Get the correct product object from the provided request."""
+
+    family = _get_family(request)
+    product_class = _get_implementing_class(family)
+    product = product_class.from_request(request)
+    product.family = family
```

Comment:
Doing this outside makes the `from_request` stuff much easier, as only the `GenericProduct` should really have a variable value and it meant passing it through unnecessarily.